### ISSUE TITLE
SXD: Benchtop Sample Imaging

### ISSUE THREAD

**Issue by KathrynBaker** [ISISComputingGroup/IBEX#4060]
As a scientist on SXD I want to be able to quickly take images of a sample before it is put onto the beamline in the cycle start in June 2019

**Acceptance Criteria**
1. There is a system available which will allow for a sample to be rotated a set number of degrees up to 360 and an image taken at each point without the need for human intervention (e.g. image at 0, move +5, take an image, move +5, take an image...)

**Notes**
1. This may need to be a 'standalone' application, rather than explicitly part of the control system
1. The motion control is undertaken via a motor driven by an Arduino developed by the Electronics team, the comms setting is 9600-8-None-1 with the following command set:

    Send | Confirmation of Receipt | Received when Done | Notes
    --- | --- | --- | ---
    `MVFWn` | `FWn` | `DONE` | Move forward n degrees
    `MVBWn` | `BWn` | `DONE` | Move backward n degress
    Anything else | `NAC` | | Not a command

1. The camera is a https://www.theimagingsource.com/products/industrial-cameras/usb-2.0-color/dfk41au02/, a USB 2.0 device, main aim would be to just take a photo once a move is complete - this still needs to be investigated
1. This appeared on the beamline unexpectedly on the 28th of February 2019, and was used for a couple of experiments by manually asking the motor to move, and then manually taking the image - this is not sustainable as a technique longer term
1. It should be possible to borrow the device in March 2019, and it should then be available until the next cycle when the automation is desired
1. This is not a directly beamline control requirement, as such we should develop this in IBEX so that we only do the work once, especially as it may not be possible to use the camera on the beamline control computer. The data for the camera is also slightly outside of the normal, and the usage and storage of that still needs to be defined

**Comment by KathrynBaker** [comment]
The ticket has been updated based on discussion with the instrument scientist.

**Comment by kjwoodsISIS** [comment]
@KathrynBaker - a couple of questions:
1. notes 2 and 7 seem inconsistent with each other.  What is the need (use-case) for a "standalone" application?
2. What does "Images from the camera should be saved alongside SXD data" mean?  Does it mean storing the images inside NeXuS files?  Or as separate files (1 file per image), each labelled with RB number, etc.?  (I am guessing it is the latter.)

**Comment by KathrynBaker** [comment]
Apologies for confusion, this has been a piece of work that was badly communicated to us as regards existence, never mind anything else.

The code has to run on a machine other than the IBEX server – standalone (note 2) because of the nature of the camera, however, if we maintain the work within IBEX (note 7) then we can guarantee the quality of the code, it will be written to our standards, and should be easily maintainable in that scenario.

Images saved alongside SXD data would indeed mean separate files alongside the NeXuS file, potentially in a directory, labelled along the lines of rbnumber-0, rbnumber-45, rbnumber-90...

**Comment by ChrisM-S** [comment]
Yes, the choice of a USB camera is not helpful when a Gig-E version would have potentially given us full control over the image capture on the NDX with area-detector (and make subsequent arching with the data much more straightforward).  Nevertheless, we do essentially have this issue on IMAT as well as the RB linking requirement.  On the IMAT cameras the imaging data is stored in RB numbered folders from the start and image taking is controlled by a script on the NDX talking to an IOC on the camera control PC.  Data synchronization is the major headache.  We need to meet between ourselves and think how best to tackle this.

**Comment by ChrisM-S** [comment]
Testing if a USB/Ethernet converter works with this camera might be useful.

**Comment by KathrynBaker** [comment]
I am changing the main ticket information as the protocol has been altered (and we weren't informed in advance of this)

**Comment by ChrisM-S** [comment]
It looks like this is likely to need to be attached to a separate control PC.  Just in terms of classifying it for support/development, it looks to be in the same category as the Triton "Device Control Computer" (#3125) - which would be an EPICS instance but without local database or IBEX GUI (The SXD instrument control machine providing the logs etc). Also, as with other imaging cameras returning data to be archived, it would need to be controlled by a script from the Instrument control machine able to ensure naming files with the current run numbers and probably sequence numbers or angles (similar to on IMAT we should enforce the "**INSTnnnnnnnn_**" prefix convention in a sub-folder of the RB number).  This would not preclude taking offline test images which are not archived, but following the camera model we would ensure that only images in RB (or XB) folder trees would be archived with the experiment (and zipped within appropriate groups).  Almost certainly this would be a fit for a "small", non-vm standard device control computer (e.g. Dell micro 7050/60).

**Comment by John-Holt-Tessella** [comment]
Can we get the protocol rewritten so that we can have a readback to know what the current position is, or if that is not possible that it has finished moving? The angle would be best because then if I restart my IOC in the middle of a run then I don't have to assume I know where it is pointing. Otherwise I can take picture to early resulting in a blurry image or assume it has moved when it hasn't resulting in a null experiment.
Is it too late to suggest they find themselves a protable Galil, like IMAT, or a McLennan, like IRIS?

We may alread have communicated this to them and they have accepted the risk but we should record that somewhere.

**Comment by ChrisM-S** [comment]
I was thinking this way initially but as it will be running from a script, it is probably just a matter of trust that the user does a sensible manual setup and uses suitable delays, e.g.

  0) BEGIN run (for indexing data files only)
  1) Line up camera manually on first sample position.
  2) start with an `angle=nnn` in the script.
  3) Increment angle (by say 10), move +ve (by 10 hopefully) - pass to local IOC (to name next file).
  4) Trigger photo - local IOC needs to poke camera software and save (area Detector image ideally).
  5) after appropriate pause, rename/save file correctly if needed, loop back and carry on.
  6) after sufficient positions END run.

Access streamed data on analysis machine if necessary, camera data archived along with other run data automatically once run has ended.

If we try to do complicated things, we risk falling into a gap where everyone will be dissapointed and spend (likely waste) a lot of time.  We simply can't do motion control 101 and build a motor record and do synchronised acquisition for something without the necessary bits.  Could be stage two to write a motor record for an Arduino instead of using a beckhoff but is this really cost efficient or where we want to be?

**Comment by ChrisM-S** [comment]
Actually, if we could use a supported M/C system from the VM and also a Gig-E camera with synchronised image taking directly to the VM, the solution might appear over engineered but we would likely save enough time and effort to make it worth our funding it! (including saving a complete separate PC & support). I think we probably should ask the question...  Gig-E cameras are not necessarily that expensive.
